cpp/src/arrow/compute/exec/benchmark_scripts/table_generation/batch_process.py

```diff
@@ -0,0 +1,104 @@
+import json
+from datagen import *
+
+from pyarrow import ipc
+from pyarrow.ipc import RecordBatchStreamWriter
+import pyarrow as pa
+import pyarrow.parquet as pq
+
+
+class BatchStreamStats:
+    def __init__(self, filename):
+        self.filename = filename
+        self.num_batches = 0
+        self.num_rows = 0
+        self.num_bytes = 0
+
+    def close(self):
+        if filename:
+            with open(filename, "w") as f:
+                f.write(f"num_batches = {self.num_batches}\n"
+                        f"num_rows = {self.num_rows}\n"
+                        f"num_bytes = {self.num_bytes}\n"
+                )
+
+    def process(self, batch_source):
+        for batch in batch_source:
+            self.num_batches += 1
+            self.num_rows += batch.num_rows
+            self.num_bytes += batch.nbytes
+
+class BatchStreamParquetFileWriter:
+    def __init__(self, filename):
+        self.filename = filename
+        self.pqwriter = None
+
+    def close(self):
+        if self.pqwriter is not None:
+            self.pqwriter.close()
+
+    def process(self, batch_source):
+        for batch in batch_source:
+            table = pa.Table.from_batches([batch], schema=batch.schema)
+            if self.pqwriter is None:
+                self.pqwriter = pq.ParquetWriter(self.filename, table.schema)
+            self.pqwriter.write_table(table)
+
+class BatchStreamFeatherFileWriter:
+    def __init__(self, filename):
+        self.filename = filename
+        self.file = None
+        self.ipcwriter = None
+
+    def close(self):
+        if self.ipcwriter is not None:
+            self.ipcwriter.close()
+        if self.file is not None:
+            self.file.close()
+
+    def process(self, batch_source):
+        if self.file is None:
+            self.file = open(self.filename, "wb")
+        for batch in batch_source:
+            if self.ipcwriter is None:
+                self.ipcwriter = ipc.new_file(self.file, batch.schema)
+            self.ipcwriter.write_batch(batch)
+
+class BatchStreamRbsWriter:
+    def __init__(self, filename):
+        self.filename = filename
+        self.file = None
+        self.rbswriter = None
+
+    def close(self):
+        if self.rbswriter is not None:
+            self.rbswriter.close()
+        if self.file is not None:
+            self.file.close()
+
+    def process(self, batch_source):
+        if self.file is None:
+            self.file = open(self.filename, "wb")
+        for batch in batch_source:
+            if self.rbswriter is None:
+                self.rbswriter = RecordBatchStreamWriter(self.file, batch.schema)
+            self.rbswriter.write_batch(batch)
```

Comment:
`pyarrow.dataset.write_dataset` accepts an iterable of record batches as its input.  Each of these classes could probably be replaced by a single call to `write_dataset`.